File: exam-back/src/main/java/com/examback/common/jwt/provider/JwtProvider.java
package com.examback.common.jwt.provider;

import com.examback.common.config.AppConfig;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import lombok.RequiredArgsConstructor;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.Map;

@Component
@RequiredArgsConstructor
public class JwtProvider {
    private static ObjectMapper getObjectMapper() {
        return AppConfig.objectMapper();
    }
    private final SecretKey jwtSecretKey;

    private SecretKey getSecretKey() {
        return jwtSecretKey;
    }

    public String generateAccessToken(Map<String, Object> claims, int seconds) {
        long now = new Date().getTime();
        Date accessTokenExpiresIn = new Date(now + 1000L * seconds);

        return Jwts.builder()
                .claim("body", jsonToStr(claims))
                .setExpiration(accessTokenExpiresIn)
                .signWith(getSecretKey(), SignatureAlgorithm.HS512)
                .compact();
    }

    public static Object jsonToStr(Map<String, Object> claims) {
        ObjectMapper objectMapper = new ObjectMapper();

        try {
            return objectMapper.writeValueAsString(claims);
        } catch (JsonProcessingException e) {
            //임시로 아무 예외 선언
            throw new DataIntegrityViolationException("예상치 못한 에러가 발생했습니다.");
        }
    }

    public boolean verify(String token) {
        try {
            Jwts.parserBuilder()
                    .setSigningKey(getSecretKey())
                    .build()
                    .parseClaimsJws(token);
        } catch (Exception e) {
            return false;
        }

        return true;
    }

    public Map<String, Object> getClaims(String token) {
        String body = Jwts.parserBuilder()
                .setSigningKey(getSecretKey())
                .build()
                .parseClaimsJws(token)
                .getBody()
                .get("body", String.class);

        return jsonToMap(body);
    }

    private Map<String, Object> jsonToMap(String jsonStr) {
        try {
            return getObjectMapper().readValue(jsonStr, LinkedHashMap.class);
        } catch(JsonProcessingException e) {
            throw new DataIntegrityViolationException("..");
        }
    }
}

File: kotlin-back/src/main/kotlin/com/example/kotlinback/global/jwt/provider/JwtProvider.kt
package com.example.kotlinback.global.jwt.provider

import com.example.kotlinback.global.config.AppConfig
import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.stereotype.Component
import java.util.*
import javax.crypto.SecretKey
import kotlin.collections.HashMap

@Component
class JwtProvider {
    private val secretKey: SecretKey? = null
    fun generateAccessToken(claims: Map<String, Any>, seconds: Int): String {
        val now = Date().time
        val accessTokenExpiresIn = Date(now + 1000L * seconds)
        return Jwts.builder()
            .claim("body", jsonToStr(claims))
            .setExpiration(accessTokenExpiresIn)
            .signWith(secretKey, SignatureAlgorithm.HS512)
            .compact()
    }

    fun verify(token: String?): Boolean {
        try {
            Jwts.parserBuilder()
                .setSigningKey(secretKey)
                .build()
                .parseClaimsJws(token)
        } catch (e: Exception) {
            return false
        }
        return true
    }

    fun getClaims(token: String?): Map<String, Any> {
        val body = Jwts.parserBuilder()
            .setSigningKey(secretKey)
            .build()
            .parseClaimsJws(token)
            .body
            .get("body", String::class.java)
        return jsonToMap(body)
    }

    private fun jsonToMap(jsonStr: String): Map<String, Any> {
        return try {
            objectMapper.readValue(jsonStr, LinkedHashMap::class.java) as Map<String, Any>
        } catch (e: JsonProcessingException) {
            throw DataIntegrityViolationException("..")
        }
    }

    companion object {
        private val objectMapper: ObjectMapper
            get() = AppConfig.objectMapper()

        fun jsonToStr(claims: Map<String, Any>): Any {
            val objectMapper = ObjectMapper()
            return try {
                objectMapper.writeValueAsString(claims)
            } catch (e: JsonProcessingException) {
                //임시로 아무 예외 선언
                throw DataIntegrityViolationException("예상치 못한 에러가 발생했습니다.")
            }
        }
    }
}
